node/src/exchange/server.rs

```diff
@@ -1,52 +1,389 @@
+use std::fmt::{Debug, Display};
+use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 
-use celestia_proto::p2p::pb::{HeaderRequest, HeaderResponse};
+use celestia_proto::p2p::pb::{header_request, HeaderRequest, HeaderResponse};
+use celestia_types::Hash;
+use futures::stream::FuturesUnordered;
+use futures::{future::BoxFuture, FutureExt, Stream};
 use libp2p::{
     request_response::{InboundFailure, RequestId, ResponseChannel},
     PeerId,
 };
-use tracing::instrument;
+use tendermint::hash::Algorithm;
+use tracing::{instrument, trace};
 
+use crate::exchange::utils::{ExtendedHeaderExt, HeaderRequestExt, HeaderResponseExt};
+use crate::exchange::{ReqRespBehaviour, ResponseType};
 use crate::store::Store;
 
-pub(super) struct ExchangeServerHandler<S>
+pub(super) struct ExchangeServerHandler<S, R = ReqRespBehaviour>
 where
-    S: Store + 'static,
+    S: Store,
+    R: ResponseSender,
 {
-    _store: Arc<S>,
+    store: Arc<S>,
+    futures: FuturesUnordered<BoxFuture<'static, (R::Channel, ResponseType)>>,
+}
+
+pub(super) trait ResponseSender {
+    type Channel: Send + 'static;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType);
 }
 
-impl<S> ExchangeServerHandler<S>
+impl ResponseSender for ReqRespBehaviour {
+    type Channel = ResponseChannel<ResponseType>;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType) {
+        // response was prepared specifically for the request, we can drop it
+        // in case of error we'll get Event::InboundFailure
+        let _ = self.send_response(channel, response);
+    }
+}
+
+impl<S, R> ExchangeServerHandler<S, R>
 where
     S: Store + 'static,
+    R: ResponseSender,
 {
     pub(super) fn new(store: Arc<S>) -> Self {
-        ExchangeServerHandler { _store: store }
+        ExchangeServerHandler {
+            store,
+            futures: FuturesUnordered::new(),
```

Comment:
I know `FuturesUnordered` suggestion was mine, but for some reason I didn't think at that time that we could just use `spawn` and an mpsc channel instead. There are some pros and cons for both cases, we can discuss them and decide.

Reply:
switched to `spawn`ing and receiving the result over channel as discussed :+1: